File: aspnet-in-practice/08-09/CS/CoolMVBlog/Utils/SEORedirectModule.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;

namespace CoolMvcBlog.Utils
{
    public class SEORedirectModule : IHttpModule
    {
        public void Dispose()
        {
        }

        public void Init(HttpApplication context)
        {
            context.BeginRequest += new EventHandler(context_BeginRequest);
        }

        private void context_BeginRequest(object sender, EventArgs e)
        {
            var context = HttpContext.Current;
            var url = context.Request.Url.AbsoluteUri;

            if (!string.IsNullOrEmpty(url.GetExtension()))
                return;

            string newUrl = url.AppendTrailingSlash();

            if (newUrl != context.Request.Url.AbsoluteUri)
                context.Response.RedirectPermanent(newUrl);
        }
    }
}

File: aspnet-in-practice/08-09/VB/CoolMVCBlog/Utils/SEORedirectModule.vb
Public Class SEORedirectModule
    Implements IHttpModule

    Public Sub Dispose() Implements IHttpModule.Dispose

    End Sub

    Public Sub Init(ByVal context As HttpApplication) Implements IHttpModule.Init
        AddHandler context.BeginRequest, AddressOf context_BeginRequest
    End Sub

    Private Sub context_BeginRequest(ByVal sender As Object, ByVal e As EventArgs)
        Dim context = HttpContext.Current
        Dim url = context.Request.Url.AbsoluteUri

        If Not String.IsNullOrEmpty(url.GetExtension()) Then
            Return
        End If

        Dim newUrl As String = url.AppendTrailingSlash()

        If newUrl <> context.Request.Url.AbsoluteUri Then
            context.Response.RedirectPermanent(newUrl)
        End If
    End Sub
End Class